ktor-server/ktor-server-plugins/ktor-server-websockets/jvm/src/io/ktor/server/websocket/WebSocketServerSession.kt

```diff
@@ -29,6 +32,68 @@ public interface DefaultWebSocketServerSession : DefaultWebSocketSession, WebSoc
  */
 public val WebSocketServerSession.application: Application get() = call.application
 
+/**
+ * Converter for web socket session
+ */
+public val WebSocketServerSession.converter: WebsocketContentConverter? get() =
+    application.plugin(WebSockets).contentConverter
+
+/**
+ * Serializes [data] of type [T] to frame using websocket content converter and enqueue this frame,
+ * may suspend if outgoing queue is full.
+ * May throw an exception if outgoing channel is already closed, so it is impossible to transfer any message.
+ * Frames that were sent after close frame could be silently ignored.
+ * Please note that close frame could be sent automatically in reply to a peer close frame unless it is
+ * raw websocket session.
+ *
+ * @throws WebsocketConverterNotFoundException if no [contentConverter] is found for the [WebSockets] plugin
+ */
+public suspend inline fun <reified T : Any> WebSocketServerSession.sendSerialized(data: T) {
+    val serializedData = converter?.serialize(
+        charset = call.request.headers.suitableCharset(),
+        typeInfo = typeInfo<T>(),
+        value = data
+    ) ?: throw WebsocketConverterNotFoundException("No converter was found for websocket")
+
+    outgoing.send(serializedData)
+}
+
+/**
+ * Dequeue frame and deserializes to type [T] using websocket content converter.
+ * May throw an exception [WebsocketDeserializeException] if converter can't deserialize frame data to type [T].
+ * May throw [WebsocketDeserializeException] if received frame type is not [Frame.Text] or [Frame.Binary].
+ * In this case [WebsocketDeserializeException.frame] contains received frame.
+ *
+ * May throw [ClosedReceiveChannelException] if channel was closed
+ *
+ * @throws WebsocketConverterNotFoundException if no [contentConverter] is found for the [WebSockets] plugin
+ * @throws WebsocketDeserializeException if received frame can't be deserialized to type [T]
+ */
+public suspend inline fun <reified T : Any> WebSocketServerSession.receiveDeserialized(): T {
```

Comment:
is it possible to avoid this code duplication in client and server?

Reply:
Moved serialization implementation to `KotlinxSerializationBase`